cpp/src/arrow/pretty_print-test.cc

```diff
@@ -64,36 +73,82 @@ void Check(const T& obj, const PrettyPrintOptions& options, const char* expected
+  static const char* expected = "[\n  0,\n  1,\n  NA,\n  3,\n  NA\n]";
+  CheckPrimitive<Int32Type, int32_t>({0, 10}, is_valid, values, expected);
+  static const char* ex_in2 = "  [\n    0,\n    1,\n    NA,\n    3,\n    NA\n  ]";
+  CheckPrimitive<Int32Type, int32_t>({2, 10}, is_valid, values, ex_in2);
+  static const char* ex_in2_w2 = "  [\n    0,\n    1,\n    ...\n    3,\n    NA\n  ]";
+  CheckPrimitive<Int32Type, int32_t>({2, 2}, is_valid, values, ex_in2_w2);
+
+  std::vector<double> values2 = {0., 1., 2., 3., 4.};
+  static const char* ex2 = "[\n  0,\n  1,\n  NA,\n  3,\n  NA\n]";
+  CheckPrimitive<DoubleType, double>({0, 10}, is_valid, values2, ex2);
+  static const char* ex2_in2 = "  [\n    0,\n    1,\n    NA,\n    3,\n    NA\n  ]";
+  CheckPrimitive<DoubleType, double>({2, 10}, is_valid, values2, ex2_in2);
+
+  std::vector<std::string> values3 = {"foo", "bar", "", "baz", ""};
+  static const char* ex3 = "[\n  \"foo\",\n  \"bar\",\n  NA,\n  \"baz\",\n  NA\n]";
+  CheckPrimitive<StringType, std::string>({0, 10}, is_valid, values3, ex3);
+  static const char* ex3_in2 =
+      "  [\n    \"foo\",\n    \"bar\",\n    NA,\n    \"baz\",\n    NA\n  ]";
+  CheckPrimitive<StringType, std::string>({2, 10}, is_valid, values3, ex3_in2);
 }
 
 TEST_F(TestPrettyPrint, BinaryType) {
   std::vector<bool> is_valid = {true, true, false, true, false};
   std::vector<std::string> values = {"foo", "bar", "", "baz", ""};
-  static const char* ex = R"expected([666F6F, 626172, null, 62617A, null])expected";
-  CheckPrimitive<BinaryType, std::string>(0, is_valid, values, ex);
+  static const char* ex = "[\n  666F6F,\n  626172,\n  NA,\n  62617A,\n  NA\n]";
+  CheckPrimitive<BinaryType, std::string>({0}, is_valid, values, ex);
+  static const char* ex_in2 =
+      "  [\n    666F6F,\n    626172,\n    NA,\n    62617A,\n    NA\n  ]";
+  CheckPrimitive<BinaryType, std::string>({2}, is_valid, values, ex_in2);
+}
+
+TEST_F(TestPrettyPrint, ListType) {
+  Int64Builder* int_builder = new Int64Builder();
+  ListBuilder list_builder(default_memory_pool(),
+                           std::unique_ptr<ArrayBuilder>(int_builder));
+
+  ASSERT_OK(list_builder.Append());
+  ASSERT_OK(int_builder->AppendNull());
+  ASSERT_OK(list_builder.Append());
+  ASSERT_OK(list_builder.Append(false));
+  ASSERT_OK(list_builder.Append());
+  ASSERT_OK(int_builder->Append(4));
+  ASSERT_OK(int_builder->Append(6));
+  ASSERT_OK(int_builder->Append(7));
+  ASSERT_OK(list_builder.Append());
+  ASSERT_OK(int_builder->Append(2));
+  ASSERT_OK(int_builder->Append(3));
+
+  std::shared_ptr<Array> array;
+  ASSERT_OK(list_builder.Finish(&array));
+  static const char* ex =
+      "[\n  [\n    NA\n  ],\n  [],\n  NA,\n  [\n    4,\n    6,\n    7\n  ],\n  [\n    "
+      "2,\n    3\n  ]\n]";
+  CheckArray(*array, {0, 10}, ex);
+  static const char* ex_2 =
+      "  [\n    [\n      NA\n    ],\n    [],\n    NA,\n    [\n      4,\n      6,\n      "
+      "7\n    ],\n    [\n      2,\n      3\n    ]\n  ]";
```

Comment:
Should we try to use the `R"` syntax for multiline output for readability (I see some used below)? No big deal here

Reply:
Yes, we should do that again. I had to remove it for some debugging but forgot re-add it again. The R-syntax is much more readable.